Readme for Nix (#1929)
o1-labs/o1js: README-nix.md

# o1js README-nix

[Nix](https://nixos.org/) is a tool for package management and system

> configurationthat can help developers build a project in a reproducible and

Reviewer: ```suggestion
configuration that can help developers build a project in a reproducible and
```

Much like the `mina` repository, you can use Nix to
handle the dependencies required across the codebase, including npm scripts.

> **When should I use Nix?**
> If you cannot build the codebase locally (due to untrusty package manager,
> faulty versioning, or unavailable libraries), it is a good idea to try the Nix
> build instead. This can specially happen if you're using a Mac–and even more
> likely–with non-Intel chips.

## Installing Nix

The following command will install Nix on your machine.

```console
sh <(curl -L https://nixos.org/nix/install) --daemon
```

If you're unsure about your Nix setup, the assistant will guide you towards a
clean installation. It will involve backing up your old `/etc/X` and `/etc/X.backup-before-nix` for `X = {bash.bashrc, bashrc, zshrc}`, and finally
uninstalling and reinstalling Nix from scratch.

> **warning for macOS users**: macOS updates will often break your nix
> installation. To prevent that, you can add the following to your `~/.bashrc`
> or `~/.zshrc`:
>
> ```bash
> # avoid macOS updates to destroy nix
> if [ -e '/nix/var/nix/profiles/default/etc/profile.d/>nix-daemon.sh' ]; then

> >  source '/nix/var/nix/profiles/default/etc/profile.d/nix-daemon.sh'

Reviewer 2: I am not sure why this is the fix, but I've seen online that OS updates will cause nix to break on macOS. I would like to try this installer in the future on a clean machine and see if it lets us avoid this step https://determinate.systems/posts/nix-survival-mode-on-macos/

Once unpacked, you should see a similar structure:

```console
rust-std-1.82.0-wasm32-unknown-unknown
├── components
├── install.sh
├── rust-installer-version
└── rust-std-wasm32-unknown-unknown
    ├── lib
    │   └── rustlib
    │       └── wasm32-unknown-unknown
```

Finally, the `wasm32-unknown-unknown` folder must be moved into the
`./lib/rustlib/` directory in the sysroot like so:

Let `{SYSROOT}` be the directory of the Rust in Nix shown in the error code;
e.g. `/nix/store/w30zw23kmgks77d870i502a3185hjycv-rust`, then the Wasm target
can be automatically installed downloading it from

```bash
mv rust-std-1.82.0-wasm32-unknown-unknown/rust-std-wasm32-unknown-unknown/lib/rustlib/wasm32-unknown-unknown {SYSROOT}/lib/rustlib/wasm32-unknown-unknown/
```

### Cargo not found

```console
error: "/nix/store/w30zw23kmgks77d870i502a3185hjycv-rust/lib/rustlib/src/rust/Cargo.lock"
       does not exist, unable to build with the standard library, try:
                   rustup component add rust-src --toolchain nix
```

#### Fix

Install `cargo` on your host machine. If doing so does not solve the problem, it

> is very likely that the whole Nix setup has gone wrong and it is very adviceable

Reviewer 3: ```suggestion
is very likely that the whole Nix setup has gone wrong and it is very advisable
```

to install it from scratch.

### Old path

When several clones of the repository are present in the system and both have
used Nix (or if it has been moved from one location to another), Nix might cache

> an old path, impossibilitating builds. For example, typing `nix develop mina`

Reviewer 3: ```suggestion
an old path, breaking builds. For example, typing `nix develop mina`
```

would complain and produce the following error:

```console
error: resolving Git reference 'master': revspec 'master' not found
```

Then, the error message would still contain old directories.

#### Fix

> Create a new environment for Nix and start from scratch.

Reviewer 3: How?
